api/environments/permissions/permissions.py

```diff
@@ -23,15 +23,19 @@ def has_object_permission(self, request, view, obj):
 class EnvironmentPermissions(BasePermission):
     def has_permission(self, request, view):
         try:
-            if view.action == "create":
+            if view.action not in ["clone", "create"]:
```

Comment:
Could we be more pythonic here? 

```python
if view.action in ("clone, "create"):
    filter_args = {
        "clone": {"api_key": request.path_info.split("/")[-3]},
        "create": {"id": request.data.get("project")}
    }[view.action]
    project = Project.objects.get(**filter_args)
    return request.user.has_project_permission("CREATE_ENVIRONMENT", project)

# return true as all users can list and specific object permissions will be handled later
return True
```

That may not work because the `request_path...` bit may throw an index error but I think something like that would be neater if we can make it work?

Reply:
I don't think it will throw an index error because we only have used this permission class for endpoints whose schema looks like this:
`...environments/<api_key>/action` right?

Reply:
My concern is that, with my suggestion, that statement will now be evaluated with we're creating an environment too, for which I think the URI is just `/api/v1/environments` ? Maybe that will be fine too, but it doesn't feel _quite_ right. Perhaps the previous code was the better option, I just didn't quite like the duplication of the `project = Project.objects.get()` statement.

Reply:
what about something like:
```
    def has_permission(self, request, view):
        def get_action_filter(action):
            if action == "create":
                return {"id": request.data.get("project")}
            if action == "clone":
                return {"environments__api_key": request.path_info.split("/")[-3]}

        if view.action in ("clone", "create"):
            try:
                project = Project.objects.get(**get_action_filter(view.action))
                return request.user.has_project_permission(
                    "CREATE_ENVIRONMENT", project
                )
            except Project.DoesNotExist:
                return False

        # return true as all users can list and specific object permissions will be handled later
        return True


```

Nested functions can be a bit hard to read though

Reply:
I don't _think_ that actually modifies the behaviour. The `request.path_info.split()` will still be evaluated when the python module gets loaded. I think the following would possibly work: 

```python
def get_project_filter_args(r):
    if view.action == "create":
        return {"environments__api_key": r.data.get("project")}
    return {"id": r.data.get("project")}

if view.action in ("clone, "create"):
    filter_args = {
        "clone": {"api_key": request.path_info.split("/")[-3]},
        "create": {"id": request.data.get("project")}
    }[view.action]
    project = Project.objects.get(**filter_args)
    return request.user.has_project_permission("CREATE_ENVIRONMENT", project)

# return true as all users can list and specific object permissions will be handled later
return True
```

BUT I think we're overcomplicating it now (my fault entirely). I think we should go back to what you had originally 😄